Update ARCHITECTURE.md (#393)
quantumblacklabs/kedro-viz: ARCHITECTURE.md

# Architecture

This document describes the high-level architecture of Kedro-Viz. It is your starting point to learn about the codebase.

For further information, see also: 

* [Kedro-Viz contributing documentation](CONTRIBUTING.md), which covers how to start development on the project
* [Kedro-Viz style guide](STYLE_GUIDE.md), which walks through our standards and recommended best practices for our codebase

## High-level Overview

Kedro-Viz is a static [React](https://reactjs.org/) web app that displays an interactive visualisation of a Kedro pipeline. It was bootstrapped with [Create-React-App](https://create-react-app.dev/). We use [Redux](https://redux.js.org/) to manage the state, and [D3](https://d3js.org/) to render the graph. The production data API is written in Python and exposes data from a Kedro project.

Kedro-Viz can exist either as:

- A standalone web app, which is [published to PyPI](https://pypi.org/project/kedro-viz/) and can be run as a Kedro plugin from the CLI

> - A React component, which is [published to `npm`](https://www.npmjs.com/package/@quantumblack/kedro-viz) and can be imported into a larger React application

Author: I think it would make sense to write it in full then, as npmjs.com, so I'll do that

> - A React component, which is [published to `npm`](https://www.npmjs.com/package/@quantumblack/kedro-viz) and can be imported into a larger React application

Author: ```suggestion
- A React component, which is [published to npmjs.com](https://www.npmjs.com/package/@quantumblack/kedro-viz) and can be imported into a larger React application
```

To allow the Kedro-Viz web app to be used as a Kedro plugin, first the JavaScript app is compiled into a static build, then it is bundled with a simple Python server and [published to PyPI](https://pypi.org/project/kedro-viz/).

## Component package/library

To publish Kedro-Viz as a React component library, it is first transpiled to the `/lib` directory with Babel. This process requires that the web worker be fully compiled (including its dependencies) with webpack, as it exists in a separate context requiring custom webpack loaders, which cannot be relied upon in an external parent application.

> When you import Kedro-Viz from `npm`, you can pass pipeline data to the component via the `data` prop:

Author: ```suggestion
When you import Kedro-Viz from npmjs.com, you can pass pipeline data to the component via the `data` prop:
```

## Synchronous data loading

> Some data source tokens instruct the app to synchronously import data, which comes direct from files in the `src` directory (such as [animals.mock.json](/src/utils/data/animals.mock.json) and [demo.mock.json](/src/utils/data/demo.mock.json)), or is generated randomly on page-load. Random data can be seeded with a 'seed' query string in the URL, to allow randomly-generated layouts to be replicated.

Reviewer: > instruct the app to synchronously import data, which comes direct from files in the `src` directory

This is a good stylistic change (avoiding a split infinitive?), but I'm not sure that it improves clarity. There's a few different ways to import data from files and I was trying to imply that I meant [ES6 imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import), which is tricky because they don't have a very specific name. I think this is a good reminder that I needed to be more clear here. Perhaps something like this below might work? What do you think?

```suggestion
Some data source tokens instruct the app to synchronously `import` [test](/src/utils/data/animals.mock.json)/[demo](/src/utils/data/demo.mock.json) data from JSON files in the `/src/utils/data` directory, or to generate it randomly on page-load. Random data can be seeded with a 'seed' query string in the URL, to allow randomly-generated layouts to be replicated.
```

Each pipeline endpoint corresponds to a different top-level pipeline. Only one top-level pipeline should be loaded at a time, so loading data from a pipeline endpoint will reset the pipeline state in the store. Each pipeline dataset contains all the data required to render the graph.

> On first page-load, the app always loads the `/api/main` endpoint first. This is the endpoint that corresponds to the 'default' pipeline. The app can load other pipelines from `/api/pipeline/<id>`. If another pipeline is saved as the active pipeline in local storage, and if it exists in the current project, then the app will load that pipeline on first page load, but it will always load the `/api/main` endpoint first, in order to check whether the active pipeline is present at that endpoint before requesting it.

Reviewer 2: the localStorage we refer to here is the `localstorage` mechanism under `window.localStorage` - I worry that calling it local storage instead of the specific name might cause confusion? 

We could re-word this to either 'Browser local storage' or as 'the `localStorage` object'?

Reviewer: I agree, this should be kept as is - if necessary we can clarify it by wrapping it in backticks to show that we mean the object: i.e. `localStorage` not localStorage

> On first page-load, the app always loads the `/api/main` endpoint first. This is the endpoint that corresponds to the 'default' pipeline. The app can load other pipelines from `/api/pipeline/<id>`. If another pipeline is saved as the active pipeline in local storage, and if it exists in the current project, then the app will load that pipeline on first page load, but it will always load the `/api/main` endpoint first, in order to check whether the active pipeline is present at that endpoint before requesting it.

Author: ```suggestion
On first page-load, the app always loads the `/api/main` endpoint first. This is the endpoint that corresponds to the 'default' pipeline. The app can load other pipelines from `/api/pipeline/<id>`. If another pipeline is saved as the active pipeline in `localStorage`, and if it exists in the current project, then the app will load that pipeline on first page load, but it will always load the `/api/main` endpoint first, in order to check whether the active pipeline is present at that endpoint before requesting it.
```

### Node API endpoints

Each node endpoint contains data required to populate the metadata panel for that node. When a user selects a node on the graph, if data for this node is not already present, then the app will request additional node data from `/api/nodes/<id>`.

> ## `localStorage`

Reviewer: Technically this is more correct, but backticked code snippets in headings look absolutely dreadful with the Github formatting! localStorage is pretty well-known to front-end devs and I think in this case it's sufficiently recognisable without the formatting?

> ## `localStorage`

Author: ```suggestion
## localStorage
```
